src/matcher/url-matcher.ts

```diff
@@ -74,9 +74,19 @@ export class UrlMatcher extends Matcher {
 		let schemeRegex = /(?:[A-Za-z][-.+A-Za-z0-9]{0,63}:(?![A-Za-z][-.+A-Za-z0-9]{0,63}:\/\/)(?!\d+\/?)(?:\/\/)?)/,  // match protocol, allow in format "http://" or "mailto:". However, do not match the first part of something like 'link:http://www.google.com' (i.e. don't match "link:"). Also, make sure we don't interpret 'google.com:8000' as if 'google.com' was a protocol here (i.e. ignore a trailing port number in this regex)
 		    wwwRegex = /(?:www\.)/,  // starting with 'www.'
 
-		    // Allow optional path, query string, and hash anchor, not ending in the following characters: "?!:,.;"
-		    // http://blog.codinghorror.com/the-problem-with-urls/
-		    urlSuffixRegex = new RegExp( '[/?#](?:[' + alphaNumericAndMarksCharsStr + '\\-+&@#/%=~_()|\'$*\\[\\]{}?!:,.;^\u2713]*[' + alphaNumericAndMarksCharsStr + '\\-+&@#/%=~_()|\'$*\\[\\]{}\u2713])?' );
+			// special characters that are permitted within the URL suffix, but not as the final character
+			latinIntraSuffixChars = /\-\+&@#\/%=~_\(\)\|\'\$\*\[\]{}\?!:,\.;\^\u2713/.source,
+			// Japanese special characters based on common punctuation defined here: https://en.wikipedia.org/wiki/Japanese_punctuation
+			// in addition to the full width variants of the latin special chars
+			japaneseIntraSuffixChars = /ー＋＆＠＃％＝〜＿｜＄￥＊｛｝（）［］【】「」『』〝〟？！：、，…	‥。・＾｡･､/.source,
+
+			// special characters that are permitted at end of URL suffix
+			latinEndOfSuffixChars = /\-\+&@#\/%=~_\(\)\|\'\$\*\[\]{}\u2713/.source,
+			japaneseEndOfSuffixChars = /ー＋＆＠＃％＝〜＿｜＄￥｛｝（）［］【】「」『』〝〟/.source,
```

Comment:
Liking the breakout into these extra variables!

Kind of minor, but to remove duplication, could we build `latinIntraSuffixChars` and `japaneseIntraSuffixChars` out from `latinEndOfSuffixChars` and `japaneseEndOfSuffixChars`? Maybe something like the following?

```ts
latinIntraSuffixChars = latinEndOfSuffixChars + /\?!:,\.;\^/.source;
```

Reply:
Good commenting too :)